tokio/src/runtime/thread_pool/queue.rs

```diff
@@ -0,0 +1,550 @@
+            let n = src_tail.wrapping_sub(src_head);
+            let n = n - n / 2;
+
+            if n == 0 {
+                return 0;
+            }
+
+            if n > LOCAL_QUEUE_CAPACITY as u32 / 2 {
+                atomic::spin_loop_hint();
+                // inconsistent, try again
+                continue;
+            }
+
+            // Track CausalCell causality checks. The check is deferred until
+            // the compare_and_swap claims ownership of the tasks.
+            let mut check = CausalCheck::default();
+
+            for i in 0..n {
+                // Compute the positions
+                let src_pos = src_head.wrapping_add(i);
+                let dst_pos = dst_tail.wrapping_add(i);
+
+                // Map to slots
+                let src_idx = src_pos as usize & MASK;
+                let dst_idx = dst_pos as usize & MASK;
+
+                // Read the task
+                let (task, ch) = self.0.buffer[src_idx]
+                    .with_deferred(|ptr| unsafe { ptr::read((*ptr).as_ptr()) });
+
+                check.join(ch);
+
+                // Write the task to the new slot
+                dst.inner.buffer[dst_idx]
+                    .with_mut(|ptr| unsafe { ptr::write((*ptr).as_mut_ptr(), task) });
+            }
+
+            // Claim all of those tasks!
+            let actual = self
+                .0
+                .head
+                .compare_and_swap(src_head, src_head.wrapping_add(n), Release);
+
+            if actual == src_head {
+                check.check();
+                return n;
+            }
+
+            atomic::spin_loop_hint();
+        }
+    }
+}
+
+impl<T> Drop for Local<T> {
+    fn drop(&mut self) {
+        if !std::thread::panicking() {
+            assert!(self.pop().is_none(), "queue not empty");
+        }
+    }
+}
+
+impl<T> Inner<T> {
+    fn is_empty(&self) -> bool {
+        let head = self.head.load(Acquire);
+        let tail = self.tail.load(Acquire);
+
+        head == tail
+    }
+}
+
+impl<T: 'static> Inject<T> {
+    pub(super) fn new() -> Inject<T> {
+        Inject {
+            pointers: Mutex::new(Pointers {
+                is_closed: false,
+                head: None,
+                tail: None,
+            }),
+            len: AtomicUsize::new(0),
+            _p: PhantomData,
+        }
+    }
+
+    pub(super) fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Close the injection queue, returns `true` if the queue is open when the
+    /// transition is made.
+    pub(super) fn close(&self) -> bool {
+        let mut p = self.pointers.lock().unwrap();
+
+        if p.is_closed {
+            return false;
+        }
+
+        p.is_closed = true;
+        true
+    }
+
+    pub(super) fn is_closed(&self) -> bool {
+        self.pointers.lock().unwrap().is_closed
+    }
+
+    fn len(&self) -> usize {
+        self.len.load(Acquire)
+    }
+
+    pub(super) fn push_batch(
```

Comment:
Why is `push` guarded by `T: ScheduleSendOnly`, whereas `push_batch` is not?

Reply:
`ScheduleSendOnly` should be removed, it did not provide the necessary safety guarantees.